From 70f8f459221b31151e6fc4f451582267e1552178 Mon Sep 17 00:00:00 2001
From: Eli Huang <Eli_Huang@wiwynn.com>
Date: Fri, 2 Jul 2021 18:15:27 +0800
Subject: [PATCH] Add VR type in PSU sensor

---
 src/PSUSensorMain.cpp | 14 +++++++++-----
 1 file changed, 9 insertions(+), 5 deletions(-)

diff --git a/src/PSUSensorMain.cpp b/src/PSUSensorMain.cpp
index f587563..603e693 100644
--- a/src/PSUSensorMain.cpp
+++ b/src/PSUSensorMain.cpp
@@ -51,7 +51,9 @@ static std::vector<std::string> sensorTypes = {
     "xyz.openbmc_project.Configuration.ISL68137",
     "xyz.openbmc_project.Configuration.ISL68220",
     "xyz.openbmc_project.Configuration.ISL68223",
+    "xyz.openbmc_project.Configuration.ISL69242",
     "xyz.openbmc_project.Configuration.ISL69243",
+    "xyz.openbmc_project.Configuration.ISL69247",
     "xyz.openbmc_project.Configuration.ISL69260",
     "xyz.openbmc_project.Configuration.LM25066",
     "xyz.openbmc_project.Configuration.MAX16601",
@@ -63,16 +65,18 @@ static std::vector<std::string> sensorTypes = {
     "xyz.openbmc_project.Configuration.pmbus",
     "xyz.openbmc_project.Configuration.PXE1610",
     "xyz.openbmc_project.Configuration.RAA228000",
+    "xyz.openbmc_project.Configuration.RAA228006",
     "xyz.openbmc_project.Configuration.RAA228228",
     "xyz.openbmc_project.Configuration.RAA229004",
     "xyz.openbmc_project.Configuration.TPS546D24"};
 
 static std::vector<std::string> pmbusNames = {
-    "adm1272",  "adm1275",   "adm1278",   "dps800",    "ina219",
-    "ina230",   "ipsps1",    "isl68137",  "isl68220",  "isl68223",
-    "isl69243", "isl69260",  "lm25066",   "max16601",  "max20710",
-    "max20730", "max20734",  "max20796",  "max34451",  "pmbus",
-    "pxe1610",  "raa228000", "raa228228", "raa229004", "tps546d24"};
+    "adm1272",  "adm1275",   "adm1278",   "dps800", "ina219",
+    "ina230",   "ipsps1",    "isl68137",  "isl68220", "isl68223",
+    "isl69242", "isl69243", "isl69247", "isl69260", "lm25066",
+    "max16601",  "max20710", "max20730", "max20734", "max20796",
+    "max34451",  "pmbus", "pxe1610", "raa228000", "raa228006",
+    "raa228228", "raa229004", "tps546d24"};
 
 namespace fs = std::filesystem;
 
-- 
2.32.0

